src/workerd/api/streams/writable-sink.c++

```diff
@@ -0,0 +1,275 @@
+      }
+      KJ_CASE_ONEOF(errored, kj::Exception) {
+        kj::throwFatalException(kj::cp(errored));
+      }
+    }
+    KJ_UNREACHABLE;
+  }
+
+  void abort(kj::Exception reason) override final {
+    canceler.cancel(kj::cp(reason));
+    setErrored(kj::mv(reason));
+  }
+
+  rpc::StreamEncoding disownEncodingResponsibility() override final {
+    auto prev = encoding;
+    encoding = rpc::StreamEncoding::IDENTITY;
+    return prev;
+  }
+
+  rpc::StreamEncoding getEncoding() override final {
+    return encoding;
+  }
+
+ protected:
+  virtual kj::AsyncOutputStream& prepareWrite(kj::Own<kj::AsyncOutputStream>& inner) {
+    return *inner;
+  };
+
+  virtual kj::Promise<void> encodeAndWrite(
+      kj::AsyncOutputStream& output, kj::ArrayPtr<const kj::byte> data) {
+    co_await output.write(data);
+  }
+
+  virtual kj::Promise<void> encodeAndWrite(
+      kj::AsyncOutputStream& output, kj::ArrayPtr<const kj::ArrayPtr<const kj::byte>> pieces) {
+    co_await output.write(pieces);
+  }
+
+  virtual kj::Promise<void> flush(kj::AsyncOutputStream& output) {
+    // When using the default implementation, we assume IDENTITY encoding.
+    KJ_ASSERT(encoding == rpc::StreamEncoding::IDENTITY);
+    // By default there's nothing to flush.
+    co_return;
+  }
+
+  void setClosed() {
+    state.init<Closed>();
+  }
+
+  void setErrored(kj::Exception&& ex) {
+    state = kj::cp(ex);
+  }
+
+  kj::AsyncOutputStream& setStream(kj::Own<kj::AsyncOutputStream> inner) {
+    auto& ret = *inner;
+    state = kj::mv(inner);
+    return ret;
+  }
+
+  kj::OneOf<kj::Own<kj::AsyncOutputStream>, Closed, kj::Exception>& getState() {
+    return state;
+  }
+
+ private:
+  kj::OneOf<kj::Own<kj::AsyncOutputStream>, Closed, kj::Exception> state;
+  rpc::StreamEncoding encoding;
+  kj::Canceler canceler;
+};
+
+// A wrapper around a native `kj::AsyncOutputStream` which knows the underlying encoding of the
+// stream and optimizes pumps from `EncodedAsyncInputStream`.
+//
+// The inner will be held on to right up until either end() or abort() is called.
+// This is important because some AsyncOutputStream implementations perform cleanup
+// operations equivalent to end() in their destructors (for instance HttpChunkedEntityWriter).
+// If we wait to clear the kj::Own when the EncodedAsyncOutputStream is destroyed, and the
+// EncodedAsyncOutputStream is owned (for instance) by an IoOwn, then the lifetime of the
+// inner may be extended past when it should. Eventually, kj::AsyncOutputStream should
+// probably have a distinct end() method of its own that we can defer to, but until it
+// does, it is important for us to release it as soon as end() or abort() are called.
+class EncodedAsyncOutputStream final: public WritableStreamSinkImpl {
+ public:
+  explicit EncodedAsyncOutputStream(
+      kj::Own<kj::AsyncOutputStream> inner, rpc::StreamEncoding encoding)
+      : WritableStreamSinkImpl(kj::mv(inner), encoding) {}
+
+  kj::Promise<void> encodeAndWrite(
```

Comment:
These `encodeAndWrite()` implementations are identical to the default ones in the base class. Should we just delete them?

Reply:
I meant just delete `encodeAndWrite()` entirely. Their call sites could be replaced with `prepareWrite(kj::mv(inner)).write(buffer)`.